dedalus/core/problems.py

```diff
@@ -45,281 +33,134 @@
-
-    def copy(self):
-        """Copy entire namespace."""
-        copy = Namespace()
-        copy.update(self)
-        copy.allow_overwrites = self.allow_overwrites
-        return copy
-
-    def add_substitutions(self, substitutions):
-        """Parse substitutions in current namespace before adding to self."""
-        for call, result in substitutions.items():
-            # Convert function calls to lambda expressions
-            head, func = parsing.lambdify_functions(call, result)
-            # Evaluate in current namespace
-            self[head] = eval(func, self)
-
-
-# class Equation:
-
-#     def __init__(self, eq_string, condition):
-#         self.eq_string = eq_string
-#         self.condition = condition
-
-#         self.LHS_string
-#         self.RHS_string
-#         self.LHS_object
-#         self.RHS_object
-
-
-    # @property
-    # def bases(self):
-    #     return self.LHS.bases
-
-    # @property
-    # def subdomain(self):
-    #     return self.LHS.subdomain
-
-    # @property
-    # def separability(self):
-    #     return self.LHS.separability
-
-    # def check_condition(self, group_dict):
-    #     pass
-
-
-
 class ProblemBase:
-    """
-    Base class for problems consisting of a system of PDEs, constraints, and
-    boundary conditions.
-
-    Parameters
-    ----------
-    domain : domain object
-        Problem domain
-    variables : list of str
-        List of variable names, e.g. ['u', 'v', 'w']
-
-    Attributes
-    ----------
-    parameters : OrderedDict
-        External parameters used in the equations, and held constant during integration.
-    substitutions : OrderedDict
-        String-substitutions to be used in parsing.
-
-    Notes
-    -----
-    Equations are entered as strings of the form "LHS = RHS", where the
-    left-hand side contains terms that are linear in the dependent variables
-    (and will be parsed into a sparse matrix system), and the right-hand side
-    contains terms that are non-linear (and will be parsed into operator trees).
-
-    The specified axes (from domain), variables, parameters, and substitutions
-    are recognized by the parser, along with the built-in operators, which
-    include spatial derivatives (of the form "dx()" for an axis named "x") and
-    basic mathematical operators (trigonometric and logarithmic).
-
-    The LHS terms must be linear in the specified variables and first-order in
-    coupled derivatives.
-
-    """
+    """Base class for all problem types."""
 
     def __init__(self, variables, namespace=None):
         self.variables = variables
         self.LHS_variables = variables
         self.dist = unify_attributes(variables, 'dist')
         self.equations = self.eqs = []
         # Build namespace via chainmap to react to upstream changes
-        # Priority: local dict, specified namespace, parseables
+        # Priority: local namespace, external namespace, parseables
+        self.local_namespace = {}
```

Comment:
Why do we make a local_namespace? Is this so someone could modify this after the problem instantiation? I don't see this used anywhere right now.

Reply:
This is so that you can directly add things to the namespace after instantiating the problem without polluting locals, i.e. `problem.namespace['a'] = A` will make `a` available for equation parsing, but won't create/change `a` in the script namespace.